index.js

```diff
@@ -223,18 +226,33 @@ class Mapeo extends events.EventEmitter {
       cb = opts
       opts = {}
     }
+    return pump(this.createDataStream(opts), fs.createWriteStream(filename), cb)
+  }
+
+  createDataStream (opts = {}) {
     if (!opts.format) opts.format = 'geojson'
-    var GeoJSONStream = exportGeoJson(this.osm, opts)
+    var bbox = opts.bbox || [ -Infinity, -Infinity, Infinity, Infinity ]
+    var filterFn = opts.filter ? compileFilter(opts.filter) : noop
+    var osmReadStream = this.osm.query(bbox, opts)
+    var filterStream = throughFilter.obj(filterFn)
+    var geoJSONStream = exportGeoJson(this.osm, opts)
+    var outputStream = duplexify()
+    outputStream.setWritable(null)
     switch (opts.format) {
-      case 'geojson': return pump(GeoJSONStream, fs.createWriteStream(filename), cb)
+      case 'geojson':
+        outputStream.setReadable(geoJSONStream)
+        break
       case 'shapefile':
-        GeoJSONStream.pipe(concat((geojson) => {
-          var zipStream = shapefile.zipStream(JSON.parse(geojson))
-          pump(zipStream, fs.createWriteStream(filename), cb)
+        geoJSONStream.pipe(concat((geojson) => {
+          outputStream.setReadable(shapefile.zipStream(JSON.parse(geojson)))
         }))
-        return
-      default: return cb(new Error('Extension not supported'))
+        break
+      default: throw new Error('Unsupported format, must be either `geojson` or `shapefile`.')
```

Comment:
If this instead returned a readable stream that immediately produced this error on next tick, then whoever consumes this API doesn't need to remember to do a try/catch as well.

Reply:
Agreed. This kind of thing is much nicer when using (Typescript|Flow) because in most cases this is a static error by a developer calling this function with the wrong arguments, and the typing would catch that statically.